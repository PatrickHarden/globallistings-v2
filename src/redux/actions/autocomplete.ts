import { createStandardAction } from 'typesafe-actions';
import { State } from '../../types/state';
import { autocompleteSelector } from '../selectors/state';
import { updatePath } from './navigation';

export const selectedAutocompletionItem = createStandardAction('SELECT_AUTOCOMPLETION_ITEM')<
  string
>();

export const updateUrlWithSelectedLocation = (placeId: string) => (
  dispatch: Function,
  getState: () => State
) => {
  const place = autocompleteSelector(getState())[placeId];
  if (!!place) {
    dispatch(updatePath('location', place.description));
  }
};
