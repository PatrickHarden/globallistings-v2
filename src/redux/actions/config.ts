import axios from 'axios';
import {
  createRequestPending,
  createRequestFailure,
  createRequestSuccess,
} from '../../types/common/action-creators-singular';
import { Config } from '../../types/config';
import { State } from '../../types/state';
import { configEndpointForCurrentUsageTypeSelector } from '../selectors/plp';

export const createFetchConfigRequest = createRequestPending('FETCH_CONFIG_REQUEST');
export const createFetchConfigFailure = createRequestFailure('FETCH_CONFIG_FAILURE');
export const createFetchConfigSuccess = createRequestSuccess<Config, 'FETCH_CONFIG_SUCCESS'>(
  'FETCH_CONFIG_SUCCESS'
);

// This will be used in the future to switch between multiple configs
export const fetchConfig = () => (dispatch: Function, getState: () => State) => {
  const state = getState();
  const configEndpoint = configEndpointForCurrentUsageTypeSelector(state);

  dispatch(createFetchConfigRequest());
  return axios
    .get(configEndpoint)
    .then(res => {
      dispatch(createFetchConfigSuccess(res.data));
    })
    .catch(ex => {
      dispatch(createFetchConfigFailure(ex));
    });
};
