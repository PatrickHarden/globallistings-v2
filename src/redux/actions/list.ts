import { createStandardAction } from 'typesafe-actions';

export const onPropertyEnter = createStandardAction('LIST_PROPERTY_MOUSEENTER')<string>();

export const onPropertyLeave = createStandardAction('LIST_PROPERTY_MOUSELEAVE')<string>();
