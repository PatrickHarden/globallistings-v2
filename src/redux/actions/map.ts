import { createStandardAction } from 'typesafe-actions';

export const onPinClicked = createStandardAction('MAP_PIN_CLICKED')<string>();
