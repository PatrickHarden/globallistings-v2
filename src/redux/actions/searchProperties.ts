import axios from 'axios';
import { createStandardAction } from 'typesafe-actions';

import { normaliseApiResponse } from '../../api/normalisers';
import { State } from '../../types/state';
import { NormalisedSearchResult } from '../../types/results';
import {
  createRequestPending,
  createRequestFailure,
  createRequestSuccess,
} from '../../types/common/action-creators';
import {} from '../selectors/property-search-results';
import {
  propertiesSearchRequestQuerySelector,
  propertySearchRequestQuerySelector,
} from '../selectors/search-api-request';
import { plpQueryHashSelector } from '../selectors/plp';
import { stringify } from 'query-string';
import { propertyIdSelector, propertyDetailsSelector } from '../selectors/pdp';
import { docToProperty } from '../../api/normalisers/doc-to-property-mapper';
import { Property } from '../../types/property';
import { Dictionary } from '../../types/common/dictionary';

export const searchPropertiesSuccess = createRequestSuccess<
  NormalisedSearchResult,
  'SEARCH_PROPERTIES_SUCCESS'
>('SEARCH_PROPERTIES_SUCCESS');
export const searchPropertiesRequest = createRequestPending('SEARCH_PROPERTIES_REQUEST');
export const searchPropertiesFailure = createRequestFailure('SEARCH_PROPERTIES_FAILURE');

export const fetchProperties = () => (dispatch: Function, getState: () => State) => {
  const state = getState();
  const params = propertiesSearchRequestQuerySelector(state);
  const hash = plpQueryHashSelector(state);
  dispatch(searchPropertiesRequest(hash));

  return axios
    .get('/api/propertylistings/query', {
      params,
      paramsSerializer: stringify,
    })
    .then(res => {
      // TODO Update to dynamic locale once we have that setup
      const normalisedData = normaliseApiResponse('en-GB')(res.data);
      dispatch(searchPropertiesSuccess(normalisedData, hash));
    })
    .catch(ex => {
      dispatch(searchPropertiesFailure(ex, hash));
    });
};

export const searchPropertyRequest = createStandardAction('SEARCH_PROPERTY_REQUEST');
export const searchPropertySuccess = createStandardAction('SEARCH_PROPERTY_SUCCESS')<
  Dictionary<Property>
>();
export const searchPropertyFailure = createStandardAction('SEARCH_PROPERTY_FAILURE')<Error>();

export const fetchPropertyIfRequired = () => (dispatch: Function, getState: () => State) => {
  const state = getState();
  const propertyId = propertyIdSelector(state);
  const params = propertySearchRequestQuerySelector(state);
  const isPropertyDetailsLoaded = propertyDetailsSelector(state);
  if (isPropertyDetailsLoaded) {
    return;
  }

  dispatch(searchPropertyRequest());

  return axios
    .get(`/api/propertylisting/${propertyId}`, { params })
    .then(res => {
      // TODO Update to dynamic locale once we have that setup
      const normalisedData = docToProperty('en-GB')(res.data.Document);
      if (!normalisedData) {
        const err = Error('Failed mapping property');
        return dispatch(searchPropertyFailure(err));
      }
      dispatch(searchPropertySuccess({ [propertyId]: normalisedData }));
    })
    .catch(ex => {
      dispatch(searchPropertyFailure(ex));
    });
};
