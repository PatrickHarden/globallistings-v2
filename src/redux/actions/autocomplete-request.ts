import { placesAutocomplete } from '../../api/google-maps/places';
import { State } from '../../types/state';
import { placeAutoCompleteConfig } from '../selectors/config/autocomplete-config';
import {
  createRequestPending,
  createRequestFailure,
  createRequestSuccess,
} from '../../types/common/action-creators';
import { normaliseAutocompleteResponse } from '../../api/normalisers';
import { NormalisedAutocompleteResult } from '../../types/autocomplete';

export const suggestionsRequest = createRequestPending('GET_SUGGESTIONS_REQUEST');
export const suggestionsFailure = createRequestFailure('GET_SUGGESTIONS_FAILURE');
export const suggestionsSuccess = createRequestSuccess<
  NormalisedAutocompleteResult,
  'GET_SUGGESTIONS_SUCCESS'
>('GET_SUGGESTIONS_SUCCESS');

export const lookupPlaceByName = (name: string) => (dispatch: Function, getState: () => State) => {
  dispatch(suggestionsRequest(name));
  const state = getState();
  const config = placeAutoCompleteConfig(state);

  return placesAutocomplete(name, config)
    .then(suggestions => {
      const normalisedData = normaliseAutocompleteResponse(suggestions);
      dispatch(suggestionsSuccess(normalisedData, name));
    })
    .catch(ex => {
      dispatch(suggestionsFailure(ex, name));
    });
};
