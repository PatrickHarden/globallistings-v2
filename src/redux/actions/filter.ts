import { FilterModifier } from '../../types/config';
import { updateQueryParams, updatePath } from './navigation';
import { isPlpPathValue } from '../../types/plp';

export const handleFilterChange = (param: string, value: string, modifier: FilterModifier) => (
  dispatch: Function
) => {
  if (modifier === FilterModifier.url) {
    if (!isPlpPathValue(param)) {
      throw new Error(
        `Can only modify path for predefined values. Value ${param} is not supported`
      );
    }
    dispatch(updatePath(param, value));
  } else {
    dispatch(updateQueryParams({ [param]: value }));
  }
};
