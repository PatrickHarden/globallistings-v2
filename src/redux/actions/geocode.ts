import { placeGeocode } from '../../api/google-maps/places';
import { State, Geocode } from '../../types/state';
import { restrictionsSelector } from '../selectors/config/config';

import { plpLocationNameSelector } from '../selectors/location';
import {
  createRequestPending,
  createRequestFailure,
  createRequestSuccess,
} from '../../types/common/action-creators';

export const createGeocodeRequest = createRequestPending('GEOCODE_REQUEST');
export const createGeocodeFailure = createRequestFailure('GEOCODE_FAILURE');
export const createGeocodeSuccess = createRequestSuccess<Geocode, 'GEOCODE_SUCCESS'>(
  'GEOCODE_SUCCESS'
);

export const validateUrl = () => (dispatch: Function, getState: () => State) => {
  const state = getState();
  const placeName = plpLocationNameSelector(state);
  const restrictions = restrictionsSelector(state);

  dispatch(createGeocodeRequest(placeName));

  return placeGeocode(placeName, restrictions)
    .then(place => {
      dispatch(createGeocodeSuccess(place, placeName));
    })
    .catch(err => dispatch(createGeocodeFailure(err, placeName)));
};
