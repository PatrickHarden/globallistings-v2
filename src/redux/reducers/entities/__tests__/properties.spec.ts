import reducer from '../properties';
import { NormalisedSearchResult } from '../../../../types/results.js';
import { DeepPartial } from '../../../../types/common/deep-partial';
import { ActionType } from 'typesafe-actions';
import { searchPropertiesSuccess } from '../../../actions/searchProperties';

const MOCK_HASH = 'MOCK_HASH';
const CREATED_AT_MOCK = 123123123;

describe('handle SUCCESS action', () => {
  const p1: DeepPartial<NormalisedSearchResult> = {
    entities: {
      properties: {
        P1: {
          numberOfBedrooms: 1,
        },
      },
    },
    result: ['P1'],
  };

  const p2: DeepPartial<NormalisedSearchResult> = {
    entities: {
      properties: {
        P2: {
          numberOfBedrooms: 2,
        },
      },
    },
    result: ['P2'],
  };

  const p1Modified: DeepPartial<NormalisedSearchResult> = {
    entities: {
      properties: {
        P1: {
          numberOfBedrooms: 3,
        },
      },
    },
    result: ['P1'],
  };

  it('should correctly handle single SUCCESS', () => {
    const successAction = {
      payload: p1 as NormalisedSearchResult,
      type: 'SEARCH_PROPERTIES_SUCCESS',
      meta: { hash: MOCK_HASH, receivedAt: CREATED_AT_MOCK },
    } as ActionType<typeof searchPropertiesSuccess>;
    const newState = reducer({}, successAction as any);

    expect(newState).toMatchSnapshot();
  });

  it('should amend new properties', () => {
    const successAction1 = {
      payload: p1 as NormalisedSearchResult,
      type: 'SEARCH_PROPERTIES_SUCCESS',
      meta: { hash: MOCK_HASH, receivedAt: CREATED_AT_MOCK },
    } as ActionType<typeof searchPropertiesSuccess>;
    const successAction2 = {
      payload: p2 as NormalisedSearchResult,
      type: 'SEARCH_PROPERTIES_SUCCESS',
      meta: { hash: MOCK_HASH, receivedAt: CREATED_AT_MOCK },
    } as ActionType<typeof searchPropertiesSuccess>;
    const newState1 = reducer({}, successAction1 as any);
    const newState2 = reducer(newState1, successAction2 as any);

    expect(newState2).toMatchSnapshot();
  });

  it('should overwrite old properties with same id', () => {
    const successAction1 = {
      payload: p1 as NormalisedSearchResult,
      type: 'SEARCH_PROPERTIES_SUCCESS',
      meta: { hash: MOCK_HASH, receivedAt: CREATED_AT_MOCK },
    } as ActionType<typeof searchPropertiesSuccess>;
    const successAction2 = {
      payload: p1Modified as NormalisedSearchResult,
      type: 'SEARCH_PROPERTIES_SUCCESS',
      meta: { hash: MOCK_HASH, receivedAt: CREATED_AT_MOCK },
    } as ActionType<typeof searchPropertiesSuccess>;
    const newState1 = reducer({}, successAction1 as any);
    const newState2 = reducer(newState1, successAction2 as any);

    expect(newState2).toMatchSnapshot();
  });
});
