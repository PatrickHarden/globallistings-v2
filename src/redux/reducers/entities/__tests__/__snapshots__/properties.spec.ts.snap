// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handle SUCCESS action should amend new properties 1`] = `
Object {
  "P1": Object {
    "numberOfBedrooms": 1,
  },
  "P2": Object {
    "numberOfBedrooms": 2,
  },
}
`;

exports[`handle SUCCESS action should correctly handle single SUCCESS 1`] = `
Object {
  "P1": Object {
    "numberOfBedrooms": 1,
  },
}
`;

exports[`handle SUCCESS action should overwrite old properties with same id 1`] = `
Object {
  "P1": Object {
    "numberOfBedrooms": 3,
  },
}
`;
